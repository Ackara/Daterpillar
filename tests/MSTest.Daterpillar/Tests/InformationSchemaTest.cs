using Acklann.Daterpillar;
using Acklann.Daterpillar.Migration;
using ApprovalTests;
using ApprovalTests.Namers;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Shouldly;
using System;
using System.Data;
using System.Linq;
using System.Runtime.CompilerServices;

namespace MSTest.Daterpillar.Tests
{
    [TestClass]
    [DeploymentItem(FName.X86)]
    [DeploymentItem(FName.Samples)]
    [DeploymentItem(FName.daterpillarXSD)]
    public class InformationSchemaTest
    {
        private const string DBNAME = "dtpl_info_mockDb";

        public TestContext TestContext { get; set; }

        [TestMethod]
        [DataSource(SourceName.supportedDatabases)]
        public void FetchSchema_should_return_a_schema_object_retrieved_from_an_existing_sql_database()
        {
            var connectionType = (Syntax)Enum.Parse(typeof(Syntax), TestContext.DataRow[0].ToString());
            TestContext.WriteLine("context: {0}", connectionType);
            using (var connection = ConnectionFactory.CreateConnection(connectionType, DBNAME))
            {
                RunTest(connection, InformationSchemaFactory.CreateInstance(connection));
            }
        }

        #region Private Members

        private void RunTest(IDbConnection connection, IInformationSchema sut, [CallerMemberName]string caller = null)
        {
            // Act
            var localSchema = MockData.GetSchema(FName.scriptingTest_partial_schemaXML);

            connection.UseSchema(FName.scriptingTest_partial_schemaXML);
            var remoteSchema = sut.FetchSchema();
            var xml = remoteSchema.ToXml();

            bool schemasAreIdentical = Compare(localSchema, remoteSchema);

            // Assert
            schemasAreIdentical.ShouldBeTrue();
            using (ApprovalResults.ForScenario(sut.GetType().Name))
            {
                Approvals.VerifyXml(xml);
            }
        }

        private bool Compare(Schema left, Schema right)
        {
            left.Tables = left.Tables.OrderBy(x => x.Name).ToList();
            right.Tables = left.Tables.OrderBy(x => x.Name).ToList();

            for (int t = 0; t < left.Tables.Count; t++)
            {
                if (left.Tables.Count != right.Tables.Count) return false;
                if (left.Tables[t].Name != right.Tables[t].Name) return false;
                if (left.Tables[t].Columns.Count != right.Tables[t].Columns.Count) return false;
                if (left.Tables[t].Indexes.Count != right.Tables[t].Indexes.Count) return false;
                if (left.Tables[t].ForeignKeys.Count != right.Tables[t].ForeignKeys.Count) return false;

                left.Tables[t].Indexes = left.Tables[t].Indexes.OrderBy(x => x.Name).ToList();
                right.Tables[t].Indexes = right.Tables[t].Indexes.OrderBy(x => x.Name).ToList();

                for (int i = 0; i < left.Tables[t].Indexes.Count; i++)
                {
                    if (left.Tables[t].Indexes[i].Name != right.Tables[t].Indexes[i].Name) return false;
                }

                left.Tables[t].ForeignKeys = left.Tables[t].ForeignKeys.OrderBy(x => x.Name).ToList();
                right.Tables[t].ForeignKeys = right.Tables[t].ForeignKeys.OrderBy(x => x.Name).ToList();

                for (int f = 0; f < left.Tables[t].ForeignKeys.Count; f++)
                {
                    if (left.Tables[t].ForeignKeys[f].Name != right.Tables[t].ForeignKeys[f].Name) return false;
                }
            }

            return true;
        }

        #endregion Private Members
    }
}