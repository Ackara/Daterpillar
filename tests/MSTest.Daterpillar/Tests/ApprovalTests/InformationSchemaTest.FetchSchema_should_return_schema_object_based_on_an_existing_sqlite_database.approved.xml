<schema name="main" xmlns="http://static.acklann.com/schema/v2/daterpillar.xsd">
  <table name="card_type">
    <column name="Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <dataType scale="32" precision="0">varchar</dataType>
    </column>
    <index name="card_type_Name" type="index" unique="true">
      <columnName order="ASC">Name</columnName>
    </index>
    <index name="card_type_Id" type="index" unique="false">
      <columnName order="ASC">Id</columnName>
    </index>
  </table>
  <table name="attribute">
    <column name="Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <dataType scale="16" precision="0">varchar</dataType>
    </column>
    <index name="attribute_Name" type="index" unique="true">
      <columnName order="ASC">Name</columnName>
    </index>
    <index name="attribute_Id" type="index" unique="false">
      <columnName order="ASC">Id</columnName>
    </index>
  </table>
  <table name="card_icon">
    <column name="Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <dataType scale="16" precision="0">varchar</dataType>
    </column>
    <index name="card_icon_Name" type="index" unique="true">
      <columnName order="ASC">Name</columnName>
    </index>
    <index name="card_icon_Id" type="index" unique="false">
      <columnName order="ASC">Id</columnName>
    </index>
  </table>
  <table name="monster_type">
    <column name="Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <dataType scale="32" precision="0">varchar</dataType>
    </column>
    <index name="monster_type_Name" type="index" unique="true">
      <columnName order="ASC">Name</columnName>
    </index>
    <index name="monster_type_Id" type="index" unique="false">
      <columnName order="ASC">Id</columnName>
    </index>
  </table>
  <table name="ability">
    <column name="Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <dataType scale="32" precision="0">varchar</dataType>
    </column>
    <index name="ability_Name" type="index" unique="true">
      <columnName order="ASC">Name</columnName>
    </index>
    <index name="ability_Id" type="index" unique="false">
      <columnName order="ASC">Id</columnName>
    </index>
  </table>
  <table name="card">
    <column name="Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <dataType scale="128" precision="0">varchar</dataType>
    </column>
    <column name="Text" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">text</dataType>
    </column>
    <column name="Pendulum_Effect" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">text</dataType>
    </column>
    <column name="Level" nullable="false" autoIncrement="false" default="'0'">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Atk" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Def" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Card_Type_Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Attribute_Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Card_Icon_Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Monster_Type_Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Ability_Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <index name="card_Monster_Type_Id" type="index" unique="false">
      <columnName order="ASC">Monster_Type_Id</columnName>
    </index>
    <index name="card_Card_Icon_Id" type="index" unique="false">
      <columnName order="ASC">Card_Icon_Id</columnName>
    </index>
    <index name="card_Attribute_Id" type="index" unique="false">
      <columnName order="ASC">Attribute_Id</columnName>
    </index>
    <index name="card_Name" type="index" unique="true">
      <columnName order="ASC">Name</columnName>
    </index>
    <index name="card_Id" type="index" unique="false">
      <columnName order="ASC">Id</columnName>
    </index>
  </table>
  <table name="card_extras">
    <column name="Card_Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Rulings" nullable="true" autoIncrement="false">
      <dataType scale="0" precision="0">text</dataType>
    </column>
    <column name="Tips" nullable="true" autoIncrement="false">
      <dataType scale="0" precision="0">text</dataType>
    </column>
    <column name="Trivia" nullable="true" autoIncrement="false">
      <dataType scale="0" precision="0">text</dataType>
    </column>
    <column name="Passcode" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <index name="card_extras_Card_Id" type="index" unique="false">
      <columnName order="ASC">Card_Id</columnName>
    </index>
  </table>
  <table name="pack">
    <column name="Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Acronym" nullable="false" autoIncrement="false">
      <dataType scale="32" precision="0">varchar</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <dataType scale="128" precision="0">varchar</dataType>
    </column>
    <column name="Description" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">text</dataType>
    </column>
    <column name="Release_Date" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">date</dataType>
    </column>
    <column name="Size" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Konami_Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <index name="pack_Name" type="index" unique="true">
      <columnName order="ASC">Name</columnName>
    </index>
    <index name="pack_Id" type="index" unique="false">
      <columnName order="ASC">Id</columnName>
    </index>
  </table>
  <table name="rarity">
    <column name="Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <dataType scale="32" precision="0">varchar</dataType>
    </column>
    <column name="Code" nullable="false" autoIncrement="false">
      <dataType scale="4" precision="0">varchar</dataType>
    </column>
    <column name="Value" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <index name="rarity_Name" type="index" unique="true">
      <columnName order="ASC">Name</columnName>
    </index>
    <index name="rarity_Id" type="index" unique="false">
      <columnName order="ASC">Id</columnName>
    </index>
  </table>
  <table name="card_number">
    <column name="Pack_Acronym" nullable="false" autoIncrement="false">
      <dataType scale="32" precision="0">varchar</dataType>
    </column>
    <column name="Number" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Rarity_Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Art_Index" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Card_Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Pack_Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <index name="card_number_Pack_Acronym_and_Number_and_Rarity_Id_and_Art_Index" type="index" unique="false">
      <columnName order="ASC">Pack_Acronym</columnName>
      <columnName order="ASC">Number</columnName>
      <columnName order="ASC">Rarity_Id</columnName>
      <columnName order="ASC">Art_Index</columnName>
    </index>
    <index name="card_number_Pack_Id" type="index" unique="false">
      <columnName order="ASC">Pack_Id</columnName>
    </index>
    <index name="card_number_Card_Id" type="index" unique="false">
      <columnName order="ASC">Card_Id</columnName>
    </index>
    <index name="card_number_Rarity_Id" type="index" unique="false">
      <columnName order="ASC">Rarity_Id</columnName>
    </index>
  </table>
  <table name="effect">
    <column name="Id" nullable="false" autoIncrement="false">
      <dataType scale="0" precision="0">int</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <dataType scale="32" precision="0">varchar</dataType>
    </column>
    <index name="effect_Name" type="index" unique="true">
      <columnName order="ASC">Name</columnName>
    </index>
  </table>
</schema>